### PR TITLE
Showroom: Detect Reload Duration

### PR DESCRIPTION
**Pull request by wlerin**
This addresses #2726 , though perhaps not in the most elegant fashion. 

A method of HLSStreamWorker is monkeypatched, and thus any future improvements to that method will not propagate to this plugin, however it hasn't been meaningfully changed in 3 years so I think that is an acceptable risk.

### CODE REVIEW

**Comment by beardypig** [comment]
FYI for others, the diff of the function is 

```diff
**diff --git a/src/streamlink/stream/hls.py b/src/streamlink/stream/hls.py
index 1328d1ae..79600506 100644
--- a/src/streamlink/stream/hls.py
+++ b/src/streamlink/stream/hls.py
@@ -221,8 +221,10 @@ class HLSStreamWorker(SegmentedStreamWorker):

         self.playlist_changed = ([s.num for s in self.playlist_sequences] !=
                                  [s.num for s in sequences])
-        self.playlist_reload_time = (playlist.target_duration or
-                                     last_sequence.segment.duration)
+        self.playlist_reload_time = max(
+            first_sequence.segment.duration,
+            last_sequence.segment.duration
+        )
         self.playlist_sequences = sequences

         if not self.playlist_changed:
```

**Comment by beardypig** [comment]
I'd actually be tempted to change it to something like

```python
self.playlist_reload_time = min([s.segment.duration for s in sequences])
```
Perhaps the average duration rounded down would also be a good value. Why did you choose the `max` duration of the two @wlerin ?

File: src/streamlink/plugins/showroom.py [hunk 1]
```diff
@@ -78,6 +79,46 @@
 ))
 
 
+# base reload time on segment duration rather than TARGETDURATION,
+# as the latter is consistently wrong
+class ShowroomHLSStreamWorker(hls.HLSStreamWorker):
+    def process_sequences(self, playlist, sequences):
```
Comment (bastimeyer) [inline src/streamlink/plugins/showroom.py]: You should not rewrite the whole `process_sequences` method here. Once the parent class gets changed, your custom one can break. The `HLSStreamWorker` class is not that flexible unfortunately and needs a bit of refactoring, but for now you could move the reload time caluclation into a separate method and then just override that in your custom class, like it's been done in my Twitch low latency PR:
https://github.com/streamlink/streamlink/pull/2513/files#diff-7ddabf15fb3a46558218929a060133e0R225-R237

**Comment by wlerin** [comment]
> I'd actually be tempted to change it to something like
> 
> ```python
> self.playlist_reload_time = min([s.segment.duration for s in sequences])
> ```
> 
> Perhaps the average duration rounded down would also be a good value. Why did you choose the `max` duration of the two @wlerin ?

I chose the max duration because that's closer to what the `TARGETDURATION` is *supposed* to be. Similarly I only used the first and last segment because accessing those segments is just a local lookup (probably inconsequential difference), and because at least right now, those are only two segments available at any given time anyway.

File: src/streamlink/plugins/showroom.py (hunk 1 above)
Comment (wlerin) [reply src/streamlink/plugins/showroom.py]: Right... but is that going to be merged soon? I didn't want to touch hls.py

Comment (bastimeyer) [reply src/streamlink/plugins/showroom.py]: > but is that going to be merged soon?

The linked PR probably not, but that's why I meant that you should add the same change in your PR.

It's a reasonable change and there's no reason why your PR shouldn't get merged if you improve the basic HLS class with that necessary change. If you want to keep this PR as simple as possible, open another one with the HLS class change alone and rebase this one after the other has been merged.

**Comment by back-to** [comment]
```py
self.playlist_reload_time = max(
            first_sequence.segment.duration,
            last_sequence.segment.duration
        )

self.playlist_reload_time = 2
```

won't work

```py
self.playlist_reload_time = 1.5
```

might work

**Pull request closed without merging**